--- jabberd/jabberd.c.orig	Mon Jan  6 18:48:57 2003
+++ jabberd/jabberd.c	Mon Jan  6 20:39:46 2003
@@ -35,6 +35,9 @@
  *
  */
 
+#include <pwd.h>
+#include <sys/types.h>
+#include <grp.h>
 #include <linux/limits.h>
 #include "jabberd.h"
 
@@ -110,9 +113,12 @@
 //    int sig;
     int help, i;           /* temporary variables */
     char *cfgfile = NULL, *c, *cmd, *home = NULL;   /* strings used to load the server config */
+	char *user = NULL;
     pool cfg_pool=pool_new();
     xmlnode pidfile;
     char *pidpath;
+	uid_t uid=0, euid=0, newgid=0;
+	struct passwd *pwd;
 
     jabberd__runtime = pool_new();
 
@@ -186,6 +192,25 @@
         fprintf(stderr,"Usage:\njabberd &\n Optional Parameters:\n -c\t\t configuration file\n -D\t\tenable debug output\n -H\t\tlocation of home folder\n -v\t\tserver version\n -V\t\tserver version\n");
         exit(0);
     }
+	
+	uid=getuid();
+	euid=geteuid();
+
+	user=ghash_get(cmd__line,"u");
+	if(user)
+	{
+		if(uid!=0) 
+			log_alert(NULL, "Cannot change user.");
+		else
+		{
+			pwd=getpwnam(user);
+			if (!pwd) log_alert(NULL, "Couldn't find user %s",user);
+			if (newgid<=0) newgid=pwd->pw_gid;
+			if (setgid(newgid)) log_alert(NULL, "Couldn't change group.");
+			if (initgroups(user,newgid)) log_alert(NULL, "Couldn't init groups.");
+			if (setuid(pwd->pw_uid)) log_alert(NULL, "Couldn't change user.");
+		}
+	}
 
     if((home = ghash_get(cmd__line,"H")) == NULL)
         home = pstrdup(jabberd__runtime,HOME);
